teachopencadd/talktorials/T018_automated_cadd_pipeline/utils/helpers/pdb.py

```diff
@@ -0,0 +1,184 @@
+    """
+    output_filepath = Path(output_filepath)
+
+    pdb_file_content = pypdb.get_pdb_file(pdb_code)
+    full_filepath = Path(f"{output_filepath}.pdb")
+    with open(full_filepath, "w") as f:
+        f.write(pdb_file_content)
+    return full_filepath
+
+
+def extract_molecule_from_pdb_file(molecule_name, input_filepath, output_filepath):
+    """
+    Extract a specific molecule (i.e. the protein or a ligand)
+    from a local PDB file and save as a new PDB file in a given path.
+
+    Parameters
+    ----------
+    molecule_name : str
+        Name of the molecule to be extracted.
+        For the protein, enter 'protein'. For a ligand, enter the ligand-ID.
+    input_filepath : str or pathlib.Path
+        Local path of the original PDB file.
+    output_filepath : str or pathlib.Path
+        Local file path (including file name) to save the PDB file of the extracted molecule in.
+
+    Returns
+    -------
+    MDAnalysis.core.universe.Universe
+        Structure object of the extracted molecule.
+    """
+
+    pdb_structure = Structure.from_string(input_filepath)
+    molecule_name = f"resname {molecule_name}" if molecule_name != "protein" else molecule_name
+    extracted_structure = pdb_structure.select_atoms(molecule_name)
+    extracted_structure.write(output_filepath)
+    return extracted_structure
+
+
+def load_pdb_file_as_dataframe(pdb_file_text_content):
+    """
+    Transform the textual content of a PDB file into a dictionary of pandas DataFrames.
+
+    Parameters
+    ----------
+    pdb_file_text_content : str
+        Textual content of a PDB file as a single string.
+
+    Returns
+    -------
+    Dict of pandas.DataFrames
+        The dictionary has 4 entries with following keys: 'ATOM', 'HETATM', 'ANISOU' and 'OTHERS'.
+        Each value is a DataFrame corresponding to the specific information described by the key.
+    """
+    ppdb = PandasPdb().read_pdb_from_list(pdb_file_text_content.splitlines(True))
+    pdb_df = ppdb.df
+    return pdb_df
+
+
+def extract_info_from_pdb_file_content(pdb_file_text_content):
+    """
+    Extract some useful information from the contents of a PDB file.
+
+    Parameters
+    ----------
+    pdb_file_text_content : str
+        Textual content of a PDB file as a single string.
+
+    Returns
+    -------
+    dict
+        Dictionary of the successfully extracted information.
+        Possible keys are:
+            'structure_title' : str
+                Title of the PDB structure.
+            'name' : str
+                Name of the protein.
+            'chains' : list of str
+                List of chain-IDs of the available chains in the protein.
+            'ligands' : list of list of str
+                List of ligand information:
+                [ligand-ID, chain-ID+residue number, number of heavy atoms]
+    """
+
+    pdb_content = pdb_file_text_content.strip().split("\n")
+
+    for index in range(len(pdb_content)):
```

Comment:
Iterate over items directly. If you need the index, `enumerate` it.

Reply:
Ok I changed it (see ae6d859), but I think it's more confusing for the reader this way, because sometimes I need the index to change the element in place, and sometimes I only want to read the element. Now if I use the index for both cases, then there is no need for `enumerate`. But if I use the index when I want to change the element, and then use the other loop variable (corresponding to the element itself) when I want to read it, then I'm referencing the same element in two different ways within the same code block, which I personally find confusing:
```python
for index, line in enumerate(pdb_content):
    pdb_content[index] = pdb_content[index].split(" ", 1)
    try:
        pdb_content[index][1] = pdb_content[index][1].strip()
        if line[1][-1] == ";":
            pdb_content[index][1] = pdb_content[index][1][:-1]
        if (
            line[0] in ["TITLE", "REMARK", "COMPND"]
            and line[1][0].isdigit()
        ):
            try:
                pdb_content[index][1] = pdb_content[index][1].split(" ", 1)[1]
            except IndexError as e:
                pass
    except:
        pdb_content[index].append(" ")
```

Reply:
Ok so I tested the function, and now it doesn't return the correct output.
The problem is that when using `enumerate`, the `enumerate` object returns the index and the element in each `__next__()` call. So when you change that element and then call it by its enumerate variable, it still returns the old unchanged value. Example:
```python
lis = ["a"]
for index, value in enumerate(lis):
    print("by index, before changing: ", lis[index])
    print("by value, before changing: ",value)
    lis[index]="z"
    print("by index, after changing: ", lis[index])
    print("by value, after changing: ",value)
```
will give:
```
by index, before changing:  a
by value, before changing:  a
by index, after changing:  z
by value, after changing:  a
```

So, the code should be like this:
```python
for index, line in enumerate(pdb_content):
    pdb_content[index] = pdb_content[index].split(" ", 1)
    try:
        pdb_content[index][1] = pdb_content[index][1].strip()
        if line[1][-1] == ";":
            pdb_content[index][1] = pdb_content[index][1][:-1]
        if (
            line[0] in ["TITLE", "REMARK", "COMPND"]
            and line[1][0].isdigit()
        ):
            try:
                pdb_content[index][1] = pdb_content[index][1].split(" ", 1)[1]
            except IndexError as e:
                pass
    except:
        pdb_content[index].append(" ")
```
meaning that we have created an extra variable `line`, but never used it.
So, in this case I think `range(len())` would have been a better choice than `enumerate`.
Nevertheless, I kept the code using `enumerate` but adapted the code to not use the `line` variable (see 0b086a2).

Reply:
changed it to (see a97fab0):
```python
for index, _ in enumerate(pdb_content):
```